From c473120eef2a86b7d330e2ca8c57153044fc3aa1 Mon Sep 17 00:00:00 2001
Message-Id: <c473120eef2a86b7d330e2ca8c57153044fc3aa1.1434674229.git.chang-joon.lee@intel.com>
In-Reply-To: <097d92eb0496291ceee6c63b06ca5def58b99eb5.1434674228.git.chang-joon.lee@intel.com>
References: <097d92eb0496291ceee6c63b06ca5def58b99eb5.1434674228.git.chang-joon.lee@intel.com>
From: Sourab Gupta <sourab.gupta@intel.com>
Date: Mon, 20 Apr 2015 12:36:55 +0530
Subject: [PATCH 2/2] SQUASHME! [VPG]: drm/i915: modify removal condition for
 memtracker sysfs file entry

SQUASHME! - This patch should be squashed into the following existing patch:
	Author: Sourab Gupta <sourab.gupta@intel.com>
	Date:   Mon Aug 11 13:12:04 2014 +0530
	Change-Id: I02552d54a753fcfb94b5bdd9bc9c82702bbf514d
	MUST_REBASE [VPG]: drm/i915: Sysfs interface to get GFX
		shmem usage stats per process

This patch fixes the condition for removal of memtracker sysfs file entries,
which has to be changed now due to an upstream patch:

commit dff01de1c3cafb128344fe809cbca84606b1b65c
Author: Chris Wilson <chris@chris-wilson.co.uk>
Date:   Thu Jul 24 14:23:10 2014 +0100

	drm: Unlink dead file_priv from list of active files first

	In order to prevent external observers walking the list of open DRM
	files from seeing an invalid drm_file_private in the process of
	being torndown, the first operation we need to take is to unlink
	the drm_file_private from that list.

This change altered the drm layer to remove the file from the filelist before
the driver->preclose (which calls the function that deletes the sysfs files) is
called. Previously it was removed after the driver close.

Because the memtrack sysfs files handler uses the filelist to check that the
file is currently only used by 1 client before removing it, with this change, it
cant find it, so refuses to remove the file.
Then, when the same program tries to create a new file when it next opens the
driver, it again doesnt see it in the list, so assumes that it is not there, and
tries to create it again, which fails due to it already existing entry.

Through this patch, we modify the condition for removing sysfs file entry, with
the knowledge that the corresponding file is already removed from the drm
filelist.

Issue: GMINL-11464
Change-Id: I7a9b8bbe70c16143e1249e57434be0dc775ae651
Signed-off-by: Sourab Gupta <sourab.gupta@intel.com>
Signed-off-by: Deepak S <deepak.s@intel.com>
---
 drivers/gpu/drm/i915/i915_sysfs.c | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/drivers/gpu/drm/i915/i915_sysfs.c b/drivers/gpu/drm/i915/i915_sysfs.c
index 297dcf2..9e6f45e 100644
--- a/drivers/gpu/drm/i915/i915_sysfs.c
+++ b/drivers/gpu/drm/i915/i915_sysfs.c
@@ -1482,15 +1482,17 @@ void i915_gem_remove_sysfs_file_entry(struct drm_device *dev,
 	struct drm_i915_private *dev_priv = dev->dev_private;
 	struct drm_i915_file_private *file_priv = file->driver_priv;
 	struct drm_file *file_local;
-	int open_count = 0;
+	int open_count = 1;
 
 	if (!i915.memtrack_debug)
 		return;
 
 	/*
-	 * Check if drm file being removed is the last one for that
-	 * particular tgid. If so, remove the corresponding sysfs
-	 * file entry also
+	 * The current drm file instance is already removed from filelist at
+	 * this point.
+	 * Check if this particular drm file being removed is the last one for
+	 * that particular tgid, and no other instances for this tgid exist in
+	 * the filelist. If so, remove the corresponding sysfs file entry also.
 	 */
 	list_for_each_entry(file_local, &dev->filelist, lhead) {
 		struct drm_i915_file_private *file_priv_local =
@@ -1500,8 +1502,6 @@ void i915_gem_remove_sysfs_file_entry(struct drm_device *dev,
 			open_count++;
 	}
 
-	WARN_ON(open_count == 0);
-
 	if (open_count == 1) {
 		struct i915_gem_file_attr_priv *attr_priv;
 
-- 
1.9.1

